Move headers Part 1

kodebach opened the pull request:
> **Note** The release notes currently don't  mention the moved/changed headers. I'll write a highlight section in one of the follow up PRs. 

This PR **starts** implementing various decisions related to headers and repo structure. It is hard to define which decisions are implemented in this PR, because the changes will be split across multiple PRs for ease of reviewing.

The entire series of PRs, will implement:

- `doc/decisions/4_decided/header_include.md`
- `doc/decisions/4_decided/header_file_structure.md`
- `doc/decisions/4_decided/library_directory_structure.md`
- `doc/decisions/4_decided/library_split.md`

## What happens in this PR?

This PR is very big so here is an overview what happens by commit (range):

1. 0db575f95f7cdf6ba1d8f537c588f86f23cb7ae9 adds the script `scripts/dev/check-includes.sh`. This script checks the new rules for includes and is also executed in a new Github Actions workflow. [Link](https://github.com/ElektraInitiative/libelektra/pull/4920/files/0db575f95f7cdf6ba1d8f537c588f86f23cb7ae9)
3. 496d1ad455ffa3d5fbbd07916471bb2fb8c72644 until d226f6be27672c44bd4d754dba2c6c7fac9dffcb move the existing headers to their new locations. Commits fall into one of these groups: (1) moving/splitting headers (2) automatic refactoring (3) reformatting (4) manual fixes of things missed by automatic script [Link](https://github.com/ElektraInitiative/libelektra/pull/4920/files/496d1ad455ffa3d5fbbd07916471bb2fb8c72644..d226f6be27672c44bd4d754dba2c6c7fac9dffcb)
4. efa1537653b00485b56592997cbe5a05fa836483 until 0a65e50bce37508943e97fd1d1841fdbcae13aa0 fix various problems resulting from the previous commits. **Note**: fd1280afb35f0bd2757be5081c4c9fe170e7f9dd is a simple regex replace (using the regex from `check-includes.sh`), which adds the now required `./` to `#include "..."` lines. [Link](https://github.com/ElektraInitiative/libelektra/pull/4920/files/efa1537653b00485b56592997cbe5a05fa836483..0a65e50bce37508943e97fd1d1841fdbcae13aa0)
5. 21b8c0dafb24a95c0447991da50c49c15416fa9c until 4bf8a751d5a702d4c2e70e3294a1476f0147f507 are again like 2 [Link](https://github.com/ElektraInitiative/libelektra/pull/4920/files/21b8c0dafb24a95c0447991da50c49c15416fa9c..4bf8a751d5a702d4c2e70e3294a1476f0147f507)
6. The remaining 8e124fdbff0cd77e6aafbb375c7f6520654b6ee5 until fbeb63388a4c2e7b4461c2d3a02f265df519c016 are again mostly fixing problems from previous changes. [Link](https://github.com/ElektraInitiative/libelektra/pull/4920/files/8e124fdbff0cd77e6aafbb375c7f6520654b6ee5..fbeb63388a4c2e7b4461c2d3a02f265df519c016) The only notable exceptions are:
    - c257793018d4469e9c0b934cc9768f8a253bef6e which removed the C89 cruft from `types.h` and thereby eliminated the need for `configure_file`. It also removes `ELEKTRA_HAVE_KDB_LONG_DOUBLE` and instead accepts a `long double` that is at least as big as `double`. **Note**: This part (`long double`) could be undone, if we don't want the change.
    - fbeb63388a4c2e7b4461c2d3a02f265df519c016 fixes a bug that comes from `master`, but seemingly wasn't detected in the CI
    - 2b0e9483836c8cda0e64276349a4be8651ab2971 temporarily disables external example tests. This will be undone in one of the follow-up PRs that finish the moving of headers.

I know the PR is still very big. However, splitting it up further would only create multiple PRs that touch about the same number of files as this PR.

### Hints for reviewers

~To reviewers I suggest using GitHub's "file viewed" checkbox feature. When you start the review go through all files quickly and check off all the files where just the `#include`s changed.~

Okay... seems that doesn't quite work, at least for me GitHub lags a lot when I try to check off the files. Using [github.dev](https://github.dev/ElektraInitiative/libelektra/pull/4920) works a bit better. 

The remaining changes should be a lot more manageable to review. See https://github.com/ElektraInitiative/libelektra/pull/4920#issuecomment-1533359038 and https://github.com/ElektraInitiative/libelektra/pull/4920#issuecomment-1560064590 for a list of those files.


## What's going to happen in follow-up PRs?

There will be 2-4 follow-up PRs depending on how easy it will be to find intermediary points where everything works.

1. The `old_helper.h` file will be split up and removed. That will allow us to re-enable the external example tests.
2. `kdbprivate.h` will be split up and removed.
3. Many CMake changes that will remove the need for `include_directories` all over the place.
4. (definitely separate) The headers from bindings also need to be moved. In particular the C++ binding still uses e.g. `#include <key.hpp>`.

Most likely, 1-2 will be one PR, 3 may be separate and 4 is definitely separate (since I haven't even started that yet). I'll keep 1 & 2 as a single PR, because these once again are things that touch large numbers of files.

## Basics

<!--
These points need to be fulfilled for every PR.
-->

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which contains `_(my name)_`)
      **Please always add them to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - add a line in `doc/news/_preparation_next_release.md`
  - reformat the code with `scripts/dev/reformat-all`
  - make all unit tests pass
  - fix all memleaks
  - fix the CI itself (or rebase if already fixed)
- [ ] The PR is rebased with current master.

<!--
If you have any troubles fulfilling these criteria, please write about the trouble as comment in the PR.
We will help you, but we cannot accept PRs that do not fulfill the basics.
-->

## Checklist

<!--
For documentation fixes, spell checking, and similar none of these points below need to be checked.
-->

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation (see [Documentation Guidelines](https://www.libelektra.org/devgettingstarted/documentation))
- [ ] I fixed all affected decisions (see [Decision Process](https://www.libelektra.org/decisions/decision-process))
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://www.libelektra.org/devgettingstarted/coding))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

<!--
Reviewers should check the following.
-->

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

<!--
If you are already Elektra developer, please adjust the labels.
Otherwise, write a comment and it will be done for you.
-->

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if everything is done** and no further pushes are planned by you.

kodebach commented:
These are the file with changes other than `#include`s:

```
.github/workflows/includes.yml
.github/workflows/macOS.yml
doc/dev/data-structures.md
doc/man/man1/kdb-run-hub-zeromq.1
doc/news/_preparation_next_release.md
doc/tutorials/highlevel-bindings.md
doc/VERSION.md
doc/VISION.md
scripts/cmake/CMakeLists.txt
scripts/cmake/ElektraConfig.cmake.in
scripts/cmake/elektra.pc.in
scripts/cmake/CMakeLists.txt
scripts/completion/install-sh-completion.in
scripts/dev/check-includes.sh
scripts/dev/ci_run_tests.sh
src/bindings/cpp/include/kdbplugin.hpp
src/bindings/cpp/include/keyset.hpp
src/bindings/glib/gelektra.pc.in
src/bindings/gsettings/CMakeLists.txt
src/bindings/io/ev/elektra-io-ev.pc.in
src/bindings/io/glib/elektra-io-glib.pc.in
src/bindings/io/uv/elektra-io-uv.pc.in
src/bindings/swig/ruby/kdb.i
src/bindings/swig/common.i
src/libs/ease/conversion.c
src/libs/elektra/symbols.map
src/libs/highlevel/elektra-codegen.pc.in
src/libs/highlevel/elektra-highlevel.pc.in
src/libs/highlevel/elektra_array_value.c
src/libs/highlevel/elektra_value.c
src/libs/io/elektra-io.pc.in
src/libs/merge/elektra-merge.pc.in
src/libs/notification/elektra-notification.pc.in
src/libs/notification/notification.c
src/libs/tools/src/command.hpp
src/plugins/blockresolver/blockresolver.c
src/plugins/curlget/curlget.c
src/plugins/directoryvalue/directoryvalue_delegate.cpp
src/plugins/doc/doc.c
src/plugins/gitresolver/gitresolver.c
src/plugins/internalnotification/internalnotification.c
src/plugins/resolver/resolver.c
src/plugins/spec/spec.h
src/plugins/type/type.c
src/plugins/type/types.c
src/plugins/type/types.h
src/plugins/version/version.h
src/plugins/yajl/iterator.c
src/plugins/yajl/yajl_gen_close.c
src/plugins/yajl/yajl_gen_open.c
tests/ctest/test_changetracking.c
tests/ctest/test_conversion.c
tests/kdb/testkdb_highlevel.cpp
tests/shell/external/CMakeLists.txt
tests/shell/external/pythongen.sh
tests/shell/gen/CMakeLists.txt
.cirrus.yml
CMakeLists.txt
```

kodebach commented:
No idea why the latex build doesn't work on Cirrus, but works on Jenkins... Any help would be appreciated

kodebach commented:
> Maybe this issue ticket helps and is related to your problem [doxygen/doxygen#9567](https://github.com/doxygen/doxygen/issues/9567)

Thanks for the tip, but that's not the issue (at least not right now).

The Cirrus docs check tried to build everything, including `elektrad`. But the go bindings can't be built against this PR (since the headers changed). I changed it so that only the docs stuff is built, since it only checks the man pages anyway. (Note: could be reduced further to just the man pages)

kodebach commented:
jenkins build libelektra please

markus2330 reviewed (changes requested):
No other reviews done so far, two required. And it is quite clear why nobody reviewed yet, as it is very challenging. Please write in [doc/news/_preparation_next_release.md](https://github.com/ElektraInitiative/libelektra/pull/4920/files#diff-b92c742dc90214bd0a67696d936102a163f58069764e572553f19594aaff679a) what the goals of these changes were. What are the automatic changes? How were they done?

It would also be very helpful if you say what are the commit ranges where you manually changed something. Maybe you even need to create a separate PR for this (just for review). For me the GitHub UI was hanging several times, e.g., it was not possible to search for a file.

kodebach commented:
> No other reviews done so far, two require

Not really my fault... Can't really do more than request reviews from people. I already reduced the size of this PR, by keeping some changes from my fork for future PRs.

> And it is quite clear why nobody reviewed yet, as it is very challenging. Please write in doc/news/_preparation_next_release.md what the goals of these changes were. What are the automatic changes? How were they done?

I did all of that already in the PR description and the first comment (linked from the description). IMO information like this has no place in the release notes. People using Elektra don't care how we made a commit. Yes, the _why_ we did that should be in the release notes, but that's why I put a note at the top of the PR description (which nobody seems to read in this repo....)

markus2330 commented:
Unfortunately still no other reviews and no changes here. I described what would have improved the chance that I will take a look. There are plenty of things that can be done, we will talk about it in the next meeting.

kodebach commented:
>  I described what would have improved the chance that I will take a look.

You wanted to know, what commits where automated, etc. I told you that this is already explained in the PR description. What else do you want from me?
If you really insist I can copy the text verbatim into `doc/news/_preparation_next_release.md`, but I really don't see the point since other than reviewers of this PR, nobody will ever care about which commits were automated...

> I described what would have improved the chance that I will take a look.

Yeah, two other reviews, but I can't do anything about that...

At this point I've got to ask: Do you even want this PR? It really seems like you don't care at all, since you're seemingly not even willing to read the PR description.

> There are plenty of things that can be done, we will talk about it in the next meeting.

The next meeting is too late for me. The PR is now open for almost 15 days (and the code is unchanged for the last 14 of those), with not a single useful comment about the content. I simply cannot afford to waste even more time, waiting for any meaningful response from you.

If this PR is not merged (or at least close to it) at the time of the next meeting, I will have to go back to developing my own fork for my thesis.

atmaxinger commented on benchmarks/changetracking.c:
```diff
@@ -1,4 +1,5 @@
-#include <benchmarks.h>
+#include "./benchmarks.h"
```
Is the `./` really necessary?

kodebach replied:
See `doc/decisions/4_decided/header_include.md`. We decided to enforce this syntax, to make sure `#include "..."` is _only_ used for local files.

atmaxinger reviewed (approved):
LGTM, I think these changes make sense overall, and thanks to the split-up of the header files we're also getting a bit more guidance on where to put new function declarations.

How much of this can be automated? Could you apply your tool also to the open PRs?

kodebach commented:
> How much of this can be automated? Could you apply your tool also to the open PRs?

There are two changes in here that affect other PRs:

First, the new rules for includes enforced by `check-includes.sh`: This one is easy to fix. Almost all includes can be fixed with regex-replace using the regexes in `check-includes.sh` as a start.

Second, the moved headers. That's  not as easy. The bigger PRs #4892 and #4922 are definitely candidates for automated updates with `tree-surgeon`. But for anything with less about 10-20 affected files (*), manually updating the includes is probably easier. My script currently doesn't account for transitive includes, so you sometimes need to add new headers even after the automatic update.

(*) Depending on the PR, not all changed files will be affected. Newly added files of course need to be changed, but modified files might be fixed with a merge/rebase after this PR.

However, if this PR is merged, I will definitely help with updating all the other PRs.

kodebach commented:
Update for https://github.com/ElektraInitiative/libelektra/pull/4920#issuecomment-1552978344: I have now moved some decisions to "partially implemented". For fixing the conflicts and (possibly) updating release notes, I'll wait until the release is done, so there won't be additional conflicts.

kodebach commented:
I have created #4940 and #4941 to show what the next steps after this PR will be. IMO it could be done as one PR, but I split the final (mostly CMake-related) changes into #4941 to simplify the reviews. One or both of these PRs will contain the proper release notes for the header & library changes.

kodebach commented:
@hannes99 Please review. To make things easier, read the PR description and the comments linked from there first.

kodebach commented:
@markus2330 Any chance that we can merge this in the next few days (until end of May)?

#4464 was merged to day, so now I'll have to update this PR again. I'd really prefer, if we could merge this PR ASAP and I'll help update all the other open PRs. We a big PR like this one, every time `master` changes, it's big gamble whether there'll be complicated conflicts to solve... Much easier to update the smaller PRs one-by-one...

hannes99 reviewed (approved):
LGTM, and seems to match what was decided in the decision.

kodebach commented:
@markus2330 This PR now has two approving reviews. I have also updated it from `master` again (now includes xfconf binding). All new files have only updated `#includes` compared `master` (*). I hope you can at least take a quick look in the next days, so we can finally merge this.


(*) except for `src/bindings/xfconf/elektra-xfconf-channel.h`, where I removed an unnecessary `ELEKTRA_UNUSED` instead of adding the header for the attribute.

kodebach commented:
For anyone interested: The `tree-surgeon` tool I wrote is now publicly available at https://github.com/kodebach/tree-surgeon. There are very few tests and zero documentation, it is now public.

kodebach commented:
As it seems this PR is not getting merged into `master` immanently, I created `new-repo-layout` and changed the base branch of this PR and merged the PR into this new branch. This way I can continue with the rest of my work and all of it can get merged into `master` together at some later point.
